lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -12,10 +12,64 @@ namespace {
 
 //==-- main implementation goes here --==//
 class TCVR {
+  ExprVector expTypeStack;
+
+  bool isConst (Expr exp) {
+   return bind::isBoolConst(exp) || bind::isIntConst(exp);
+  }
+
+  bool isVar (Expr exp) {
+   return bind::isBoolVar(exp) || bind::isIntVar(exp);
+  }
+
+  bool isValue (Expr exp) {
+    return isOpX<TRUE>(exp) || isOpX<FALSE>(exp);
+  }
+
+  //returns true if children are of correct type
+  bool checkChildren (Expr exp) {
+
+    if (isOpX<ITE>(exp)) {
+      //ite (a,b,c) : a is bool, b and c same type
+      Expr c = expTypeStack.back(); 
+      expTypeStack.pop_back();
+
+      Expr b = expTypeStack.back(); 
+      expTypeStack.pop_back();
+
+      Expr a = expTypeStack.back(); 
+      expTypeStack.pop_back();
+
+      bool properTypes = (a == sort::boolTy (exp->efac())) && (b == c);
+      return properTypes;
+    }
+    //default: check that chilren expressions are of the same types as the parent
+
+    int numChildren = exp->arity();
+    Expr parentType = bind::typeOf(exp);
+    bool properTypes = true;
+
+    for (int i = 0; i < numChildren; i ++) {
+      properTypes = properTypes && expTypeStack.back () == parentType;
+      expTypeStack.pop_back();
+    }
+
+    return properTypes;
+  }
 
   /// Called after children have been visited
   Expr postVisit(Expr exp) {
-    LOG("tc", llvm::errs() << "post visiting expression: " << *exp << "\n";);
+    LOG("tc", llvm::errs() << "post visiting expression: " << *exp << "\n";); 
+
+    if (!checkChildren(exp)) {
+      LOG("tc", llvm::errs() << "improper types at:" << *exp  << "\n";);
+
+      expTypeStack.push_back(Expr());
+      return exp;
+    }
+
+      expTypeStack.push_back(bind::typeOf(exp));
```

Comment:
The stack will not work. Expressions are a DAG not a tree.
I'd want to see a more complete and documented design of the traversal your are planning to do.
This is kind of a mess right now.